spec/safe_statements_spec.rb

```diff
@@ -1756,60 +1889,99 @@ def up
 
             expect do
               test_migration.suppress_messages { test_migration.migrate(:up) }
-            end.to raise_error(PgHaMigrations::InvalidMigrationError, "Partitioned table \"foos3\" contains sub-partitions")
+            end.to raise_error(PgHaMigrations::InvalidMigrationError, "Table \"foos3\" is not a partitioned table")
           end
 
-          it "raises error when index invalid" do
-            create_range_partitioned_table(:foos3, migration_klass, with_partman: true)
+          it "raises error and leaves behind invalid index when child index creation missed" do
+            create_range_partitioned_table(:foos3, migration_klass)
+
+            ActiveRecord::Base.connection.execute(<<~SQL)
+              CREATE TABLE foos3_child1 PARTITION OF foos3
+              FOR VALUES FROM ('2020-01-01') TO ('2020-02-01');
+
+              CREATE TABLE foos3_child2 PARTITION OF foos3
+              FOR VALUES FROM ('2020-02-01') TO ('2020-03-01');
+            SQL
 
             test_migration = Class.new(migration_klass) do
               def up
                 safe_add_concurrent_partitioned_index :foos3, :updated_at
               end
             end
 
-            allow(ActiveRecord::Base.connection).to receive(:execute).and_call_original
-
-            # skip the attachment step to simulate invalid index
-            allow(ActiveRecord::Base.connection).to receive(:execute).with(/ATTACH PARTITION/)
+            allow_any_instance_of(PgHaMigrations::Table).to receive(:partitions)
+              .and_return([PgHaMigrations::Table.new("foos3_child1", "public")])
 
             expect do
               test_migration.suppress_messages { test_migration.migrate(:up) }
             end.to raise_error(PgHaMigrations::InvalidMigrationError, "Unexpected state. Parent index \"index_foos3_on_updated_at\" is invalid")
+
+            aggregate_failures do
+              %i[foos3 foos3_child1].each do |table|
+                indexes = ActiveRecord::Base.connection.indexes(table)
+
+                expect(indexes.size).to eq(1)
+                expect(indexes.first).to have_attributes(
+                  table: table,
+                  name: "index_#{table}_on_updated_at",
+                  columns: ["updated_at"],
+                  using: :btree,
+                )
+              end
+
+              expect(ActiveRecord::Base.connection.indexes(:foos3_child2)).to be_empty
+            end
           end
 
-          it "raises error when parent index name is too large" do
+          it "raises error and leaves behind invalid index when child sub-partition index creation missed" do
```

Comment:
> I'm wondering if we want to have a test or two that shows the the invalid indexes left behind in certain error cases.

This test and the one above _sort of_ addresses the concern. I think the unexpected state error combined with expectations on the existence / absence of various indexes is enough to confirm the behavior. We could also check the validity of each index that is left behind, but that would involve either using `PgHaMigrations::Index#valid?` (which is something that's already being exercised in this test via the unexpected state error) or introduce a test helper that duplicates the SQL the method executes. I don't see a ton of value there

Reply:
I think this was addressed by adding expectations like:

```ruby
expect(ActiveRecord::Base.pluck_from_sql("SELECT indexrelid::regclass FROM pg_index WHERE NOT indisvalid")).to contain_exactly(
  "index_foos3_on_updated_at",
  "index_foos3_sub_on_updated_at",
)
```